src/sage/combinat/abstract_tree.py

```diff
@@ -1156,6 +1156,23 @@ def paths(self):
                 for p in t.paths():
                     yield (i,) + p
 
+    def _require_mutable(self):
+        """
+        Internal function. Although the general usage of abstract tree is often
+        in a non-mutable way, mutable usage also exists. This would affect
+        precomputed number such as _node_number. This function clears out the
+        precomputed values when there is a mutable usage.
+
+        Values cleared:
+
+        - ``_node_number`` : number of nodes in the tree
+        """
```

Comment:
```suggestion
        Values cleared:

        - ``_node_number`` : number of nodes in the tree

        EXAMPLES::

             sage: T = ATree()
             sage: T.node_number()
             sage: T._require_mutable()
             sage: T._node_number
             Traceback (most recent call last):
             ...
             AttributeError: [message]
        """
```
This needs a doctest, something like this.

Reply:
In fact, some more work is needed, as AbstractTree can be immutable or mutable. I have added extra tests for both cases, and fixed doctests elsewhere because I am using a new error message than super.

Reply:
I don't see that trees cannot create mutable copies. The check you added is exactly the same as what `_require_mutable()` checks (there is no `is_immutable()` defined in the `combinat` folder). So what you did should have no effect. I think you need more justification for the last change.

Reply:
Ping? I need more information here.

Reply:
@tscrim In fact, the usual trees are not mutable when created by a constructor, but as they inherit ClonableArray when we want a mutable version (via `AbstractClonableTree`), we may construct a mutable version by calling the `clone()` method. The method `is_immutable()` is also present there. I agree that it may not be a very good idea to write code in `AbstractTree` that depends on something that inherits it, but it seems to pass the test. If we really want to be pedantic, we can add a check of attribute or type just before.

For some examples, please have a look at the tests involved in the commit 046d719.

Reply:
@fwjmath I don't understand what you're saying. Yes, I know how `ClonableArray` works, which is why I said we should override this method. I don't necessarily have a problem with it being in `AbstractTree`, but `AbstractClonableTree` almost certainly should be the correct spot because all (syntactically) mutable-while-cloning trees inherit from that.

However, my previous comment was about lines 1187--1188:
```python
        if self.is_immutable():
            raise ValueError("This object is immutable.")
```
which is doing the same thing the `super()._require_mutable()` does. This part needs justification if you keep it there.

Reply:
@tscrim I put a check because if I do it there just as the code you proposed, as `AbstractTree` does not inherit `ClonableArray`, its `super` does not have `_require_mutable`, so I put the extra check there. But after a second thought, I think that it is much more reasonable to put this part into `AbstractClonableTree`, which is the real place where we ask for mutability. This is done in the latest commit, in which `is_mutable` is no longer called, and I also reverted the tests back to the old error string.

Reply:
This is moot, but if it was in `AbstractTree`, the method would have never been called anyways. Yet it would be a clear code smell.

Anyways, thank you for fixing this.